src/core/inductive.ml

```diff
@@ -0,0 +1,280 @@
+    _Impl (fapp sind ts) prop
+  in
+  let p_type = create_instance_of pos sind to_find_p_type in
+
+  (* STEP 2: Create each clause according to a constructor *)
+  (* [case_of scons] creates a clause according to a constructor [scons]. *)
+  let case_of : sym -> tbox = fun scons ->
+    let rec case : tbox list -> term -> tbox = fun xs a ->
+      match Basics.get_args a with
+      | (Symb(s), l) ->
+          (* Check if the current constructor is a constructor of the current
+             inductive type. *)
+          if s == sind then
+            let l = List.map lift l in
+            prf_of_p l (app (_Symb scons) (List.rev xs))
+          else
+            fatal pos "%a is not a constructor of %a"
+              pp_symbol scons pp_symbol sind
+      | (Prod(a,b), _) ->
+          (* Take a name and the body of the term *)
+          let (x,b) = unbind_name b                in
+          (* Do the same thing until the end of the constructor's type *)
+          let b = case ((Bindlib.box_var x)::xs) b in
+          begin
+            match Basics.get_args a with
+            | (Symb(s), l) ->
+                let b =
+                  (* If the current symbol is equal to the inductive type,
+                     we create an inductive hypothesis. *)
+                  if s == sind then
+                    let l = List.map lift l in
+                    _Impl (prf_of_p l (Bindlib.box_var x)) b
+                  else
+                    b
+                in
+                let res = _Prod (lift a) (Bindlib.bind_var x b) in
+                if !log_enabled then
+                  log_ind "The clause of the constructor [%a] is %a"
+                    Print.pp_term (Symb(scons))
+                    Print.pp_term (Bindlib.unbox res);
+                res
+            | _ -> fatal pos "The type of %a is not supported" pp_symbol scons
+          end
+      | _ -> fatal pos "The type of %a is not supported" pp_symbol scons
+    in
+    case [] !(scons.sym_type)
+  in
+
+  (* STEP 3: Create the conclusion of the inductive principle *)
+  let to_create_end  : tbox list -> tbox = fun ts ->
+    let x = Bindlib.new_var mkfree "x"                 in
+    let t = Bindlib.bind_var x (prf_of_p ts (_Vari x)) in
+    _Prod (fapp sind ts) t
+  in
+  let conclusion = create_instance_of pos sind to_create_end  in
+
+  (* STEP 4: Create the inductive principle *)
+  let rec add_case l r =
+    match l with
+    | []   -> r
+    | t::q -> let t = case_of t in
+              _Impl t (add_case q r)
+  in
+  let t = add_case scons_list conclusion      in
+  let t = _Prod p_type (Bindlib.bind_var p t) in
+  if !log_enabled then
+    log_ind "The induction principle of the inductive type [%a] is %a"
+      Print.pp_term (Symb(sind)) Print.pp_term (Bindlib.unbox t);
+  Bindlib.unbox t
+
+(** [type_to_pattern t] creates a pattern which matches with the type [t]. *)
+let rec type_to_pattern : term -> p_term = fun t ->
```

Comment:
Is this function really necessary?

Reply:
See my email. It's the dual of the function `lift` if you compare the structure of `principle` and `ind_rule`

Reply:
I think that we can get rid of this function. See my comment on the generation of rules.
Anyway, if one day, you really need to implement this function, here are several remarks:
- when doing a match on a term t, always write `match unfold t` !
- you should use the helping functions P.iden, P.app, etc.
- the definition must be completed to handle all possible terms
- a Vari can be a bound variable. in this case, it must not be replaced by a P_Patt
-  only free Vari's can be replaced by a P_Patt but, in this case, they must be renamed to avoid name clashes
- as t is arbitrary, the resulting p_term may not be a valid LHS argument